src/Rings/mpoly.jl

```diff
@@ -1107,3 +1108,86 @@ function factor(f::MPolyElem)
   return Nemo.Fac(R(fS.unit), Dict(R(k) =>v for (k,v) = fS.fac))
 end
 =#
+
+################################################################################
+
+function normal_form(f::MPolyElem, I::MPolyIdeal)
+  R = parent(f)
+  R == base_ring(I) || error("base rings do not match")
+  Oscar.groebner_assure(I)
+  L = Singular.reduce(I.gb.Sx(f), I.gb.S)
+  return R(L)
+end
+
+################################################################################
+
+function normal_form(I::MPolyIdeal, J::MPolyIdeal)
+  R = base_ring(I)
+  R == base_ring(J) || error("base rings do not match")
+  Oscar.groebner_assure(J)
+  singular_assure(I)
+  singular_assure(J)
+  L = Singular.reduce(I.gens.S, J.gb.S)
+  return ideal(R, L)
+end
+
+################################################################################
+
+function Base.:*(f::MPolyElem, I::MPolyIdeal)
+  R = base_ring(I)
+  R == parent(f) || error("base rings do not match")
+  return ideal(R, f.*gens(I))
+end
+
+################################################################################
+@doc Markdown.doc"""
+    PrimaryDecomposition(I::MPolyIdeal, algo::String="GTZ")
+
+Compute a primary decomposition of the ideal `I` using the `GTZ`-algorithm by default, or `SY` if specified. The output is a dictionnary where the keys are the primary ideals, with values the corresponding prime ideal.
+"""
+function PrimaryDecomposition(I::MPolyIdeal, algo::String="GTZ")
+  R = base_ring(I)
+  singular_assure(I)
+  if algo == "GTZ"
+    L = Singular.LibPrimdec.primdecGTZ(I.gens.Sx, I.gens.S)
+  elseif algo == "SY"
+    L = Singular.LibPrimdec.primdecSY(I.gens.Sx, I.gens.S)
+  else
+    error("algorithm invalid")
+  end
+  D = Dict{Oscar.MPolyIdeal, Oscar.MPolyIdeal}()
+    for q in L
+      D[ideal(R, q[1])] = ideal(R, q[2])
+    end
+  return D
+end
+
+################################################################################
+# I don't know if there is a smarter way to check if an ideal is prime
+
+function isprime_ideal(I::MPolyIdeal)
+  R = base_ring(I)
+  D = PrimaryDecomposition(I)
+  if length(D) != 1
+    return false
+  else
+    for (q, p) in D
+      r1 = normal_form(p, q)
+      return r1 == ideal(R, [R(0)])
+    end
+  end
+end
+
+################################################################################
+
+function isprimary_ideal(I::MPolyIdeal)
+  D = PrimaryDecomposition(I)
+  return length(D) == 1
+end
+
+################################################################################
+
+function hash(I::MPolyIdeal)
```

Comment:
Rather implement `Base.hash(I::MPolyIdeal, h::UInt)` and pass it on, see `?hash`.